### PR TITLE
Implemented cbt run command

### PR DESCRIPTION
**Pull request by ChavXO**
Before "runClass" defaulted to returning "Main" so it didn't run the file. Still needs work but this a basic implementation.

### CODE REVIEW

File: stage2/DefaultBuild.scala [hunk 1]
```diff
@@ -214,7 +214,27 @@ class Build(val context: Context) extends Dependency with TriggerLoop{
     )
   }
 
-  def runClass: String = "Main"
+	def hasMainClass(file : File, classLoader : ClassLoader) : Boolean = {
```
Comment (farmdawgnation) [inline stage2/DefaultBuild.scala]: This indentation doesn't look like what's been used throughout the rest of the code. :)

File: stage2/DefaultBuild.scala [hunk 2]
```diff
@@ -214,7 +214,27 @@ class Build(val context: Context) extends Dependency with TriggerLoop{
...
+	  val className = getClassName(file)
+	  val inspectClass = classLoader.loadClass(className)
+	  inspectClass.getDeclaredMethods().map(_.getName).contains("main")
+	}
+	
+	
+	val files = compileTarget.listFiles
+	
+	val classFiles = files.filter(x =>x.toString.endsWith("class"))
```
Comment (farmdawgnation) [inline stage2/DefaultBuild.scala]: You could use the `_` wildcard here to avoid having to use a single letter variable named `x`.

`files.filter(_.toString.endsWith("class"))`

Should we also ensure this `endsWith` checks to `.class` so that things like "Myawesomeclass" wouldn't match?

File: stage2/DefaultBuild.scala [hunk 3]
```diff
@@ -214,7 +214,27 @@ class Build(val context: Context) extends Dependency with TriggerLoop{
...
+	
+	def getClassName(file : File) = file.getName.takeWhile(_ != '.')
+	
+	val hasMain = classFiles.filter(x => hasMainClass(x, classLoader))
```
Comment (farmdawgnation) [inline stage2/DefaultBuild.scala]: Ditto here with replacing `x` with `_`.

`classFiles.filter(hasMainClass(_, classLoader))`

File: stage2/DefaultBuild.scala [hunk 4]
```diff
@@ -214,7 +214,27 @@ class Build(val context: Context) extends Dependency with TriggerLoop{
...
+	
+  def runClass: String = try {
+	  getClassName(hasMain.head)
+	} catch {
+	  case e : NoSuchElementException => "Main"
```
Comment (farmdawgnation) [inline stage2/DefaultBuild.scala]: What's the effect of returning `"Main"` here? Does it try to invoke a `Main` even if it doesn't exist?

Comment (farmdawgnation) [inline stage2/DefaultBuild.scala]: Ah, this is essentially the existing behavior, isn't it?

Comment (ChavXO) [inline stage2/DefaultBuild.scala]: Yes. Ideally though you would throw some kind of error. Not sure how dev envisions errors being propagated. Slightly reminiscent of SBT it would be nice, in the case that there are multiple files with a main method, if the tool allows the user to select which main to run. Working on that though.

Comment (farmdawgnation) [inline stage2/DefaultBuild.scala]: Makes sense. :ok:

File: stage2/DefaultBuild.scala [hunk 5]
```diff
@@ -214,7 +215,47 @@ class Build(val context: Context) extends Dependency with TriggerLoop{
...
+    ${mains.mkString("\t\n")}""")
+    //FIXME: doesn't stop to read the line!!!
+    //val choice = scala.io.StdIn.readLine() 
+    
+    files(0) // defaulting to first file in the meantime
+  }
+  
+  def getClassName(file : File) = file.getName.takeWhile(_ != '.')
+	
+  def runClass: String = {
```
Comment (cvogt) [inline stage2/DefaultBuild.scala]: So the (undocumented) intention of this method was really just to determine the name of the class, that `def run` would run. See below.

I thought about if CBT should add a main class discovery mechanism and I am not sure. Maybe. In any case that logic should go into class Lib. This Build class is really just supposed to be a shallow layer on top of it and not contains major functionality as implemented here.

By the way, opening an interactive terminal seems to be non-standard in nailgun. I didn't investigate the details, but it would be good if someone gathers the knowledge how to do that, so we can also get rid of the publishSigned hack in the bash script that prevents running through nailgun in that case. Basically the problem is java.io.Console.getConsole is null.

Comment (cvogt) [inline stage2/DefaultBuild.scala]: Yeah, I think it's a good idea actually. I had problem where I set the runClass to a non existing one and then cbt would just do nothing with the below logic.

Comment (cvogt) [inline stage2/DefaultBuild.scala]: let's add it as a method to the library and figure out how interactive works.

**Comment by cvogt** [comment]
this needs rebase on master and eventually before merging - squashing

File: stage2/DefaultBuild.scala [hunk 6]
```diff
@@ -214,7 +215,47 @@ class Build(val context: Context) extends Dependency with TriggerLoop{
...
+    if (compileTarget == null) return "Main"
```
Comment (cvogt) [inline stage2/DefaultBuild.scala]: why would it be null?

File: stage2/DefaultBuild.scala [hunk 7]
```diff
@@ -214,7 +215,47 @@ class Build(val context: Context) extends Dependency with TriggerLoop{
...
+    val hasMain = getClassFiles(compileTarget)
+    
+	  val classFile = if (hasMain.length == 1) hasMain(0) else pickMain(hasMain) 
+	  try {
+      getClassName(classFile)
+    } catch {
+      case e : NoSuchElementException => "Main"
+    }
```
Comment (cvogt) [inline stage2/DefaultBuild.scala]: this try catch seems unnecessary now, no?

File: stage2/DefaultBuild.scala (hunk 6 above)
Comment (ChavXO) [inline stage2/DefaultBuild.scala]: Trying to call cbt run before cbt compile. Ideally it should compile at that point.

Comment (cvogt) [inline stage2/DefaultBuild.scala]: so `run` calls `classLoader`so this should work when called via run. Maybe just call compile here to make sure. Also compileTarget is a File object, it shouldn't be null.

File: stage2/DefaultBuild.scala (hunk 7 above)
Comment (ChavXO) [inline stage2/DefaultBuild.scala]: Yes. Spill-Over from previous code.

File: stage2/DefaultBuild.scala [hunk 8]
```diff
@@ -214,7 +215,8 @@ class Build(val context: Context) extends Dependency with TriggerLoop{
     )
   }
 
-  def runClass: String = "Main"
+  def runClass: String = lib.getRunClass(compileTarget, classLoader)
```
Comment (cvogt) [inline stage2/DefaultBuild.scala]: let's keep the names identical like everywhere else and get rid of the `get`

File: stage2/Lib.scala [hunk 9]
```diff
@@ -65,6 +65,44 @@ final class Lib(logger: Logger) extends Stage1Lib(logger) with Scaffold{
...
+      inspectClass.getDeclaredMethods().map(_.getName).contains("main")
+    }
+
+    def isInnerClass(file : File) : Boolean = file.toString.contains("$")
+    val files = target.listFiles
+
+    val classFiles = files.filter(file => file.toString.endsWith(".class") && !isInnerClass(file)).toList
+
+    classFiles.filter(hasMainClass(_))
+	}
```
Comment (cvogt) [inline stage2/Lib.scala]: tab slipped in

File: stage2/Lib.scala [hunk 10]
```diff
@@ -65,6 +65,44 @@ final class Lib(logger: Logger) extends Stage1Lib(logger) with Scaffold{
...
+        
+    ${mains.mkString("\t\n")}""")
+
+    //FIXME: doesn't stop to read the line!!!
+    //val choice = scala.io.StdIn.readLine() 
+
+    files(0) // defaulting to first file in the meantime
+  }
+  
+  def getClassName(file : File) = file.getName.takeWhile(_ != '.')
```
Comment (cvogt) [inline stage2/Lib.scala]: how about making all of the above inner functions of getRunClass, to underline that that aren't used by anything else and keep the exported api small? If needed we can move them out again later.

File: stage2/Lib.scala [hunk 11]
```diff
@@ -65,6 +65,44 @@ final class Lib(logger: Logger) extends Stage1Lib(logger) with Scaffold{
...
+    val classFiles = files.filter(file => file.toString.endsWith(".class") && !isInnerClass(file)).toList
+
+    classFiles.filter(hasMainClass(_))
+	}
+  
+  def pickMain(files : List[File]) : File = {
+    val mains = files.map(_.getName).zipWithIndex map {case (fileName, index) => s"""[${index + 1}] ${fileName}\n"""}
+    println(s"""Multiple main classes detected. Select one to run:
+        
+    ${mains.mkString("\t\n")}""")
```
Comment (cvogt) [inline stage2/Lib.scala]: does this indent all but the first one with a tab? I think we don't need to indent them. Just scratch the \t :)

**Comment by cvogt** [comment]
How about this for a design:

**Comment by cvogt** [comment]
```
// class BasicBuild
def run: Unit = lib.run( mainClasses, Seq(), classLoader )  )
def mainClass: Option[String] = mainClasses.headOption
def mainClasses: Seq[String] = lib.mainClasses(compileTarget, classLoader)
```

```

**Comment by ChavXO** [comment]
@cvogt So  the default behaviour would be to run all main classes found in the target directory?

**Comment by cvogt** [comment]
no, only if there is exactly one. Actually the above runs simple one of them, but none in case of ambiguity is better I suppose

**Comment by cvogt** [comment]
fixing the code above

**Comment by cvogt** [comment]
.foreach is the foreach on Option there, not on Seq

**Comment by ChavXO** [comment]
@cvogt Done.

File: stage1/Stage1.scala [hunk 12]
```diff
@@ -5,6 +5,7 @@ import java.time.LocalTime.now
 
 import scala.collection.immutable.Seq
 import scala.collection.JavaConverters._
+import scala.language.postfixOps
```
Comment (cvogt) [inline stage1/Stage1.scala]: did you add this by accident?

File: stage2/BasicBuild.scala [hunk 13]
```diff
@@ -144,8 +147,16 @@ class Build(val context: Context) extends Dependency with TriggerLoop{
     )
   }
 
-  def runClass: String = "Main"
-  def run: ExitCode = lib.runMainIfFound( runClass, context.args, classLoader ) 
+  def run: ExitCode = lib.run( runClass, classLoader )
+  def runClass: Option[String] = {
+	  if (mainClasses.size == 1)
```
Comment (cvogt) [inline stage2/BasicBuild.scala]: There is a tab in here again. I suggest you find some way to let your editor know.

File: stage2/BasicBuild.scala [hunk 14]
```diff
@@ -105,6 +106,8 @@ class Build(val context: Context) extends Dependency with TriggerLoop{
   //def cacheJar = false
   override def dependencyClasspath : ClassPath = ClassPath(localJars) ++ super.dependencyClasspath
   override def dependencyJars      : Seq[File] = localJars ++ super.dependencyJars
+  
+  def clean : ExitCode = lib.clean(target)
```
Comment (cvogt) [inline stage2/BasicBuild.scala]: let's do the clean stuff in a separate PR. git cherry pick will help migrate the commit to another branch.

File: stage2/Lib.scala~ [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+            .map(_.context.toString)
+            .map(new File(_))
+          changedFiles.foreach( f => logger.loop("Changed: "+f) )
+          changedFiles.collect(action)
+          key.reset
+        }
+      }
+    }
+  }
+}
```
Comment (cvogt) [inline stage2/Lib.scala~]: This file slipped in, but shouldn't have.

File: stage2/Lib.scala [hunk 16]
```diff
@@ -66,6 +66,52 @@ final class Lib(logger: Logger) extends Stage1Lib(logger) with Scaffold{
...
+      println("[info] Failed")
+      return ExitCode.Failure
+    }
+  }
+
+  
+  def run( mainClass: Option[String], classLoader: ClassLoader ) : ExitCode = mainClass match {
+	  case Some(className) =>  runMain( className, Seq(), classLoader )
+	  case _               => ExitCode.Failure
+	}
```
Comment (cvogt) [inline stage2/Lib.scala]: tabs

Comment (cvogt) [inline stage2/Lib.scala]: also common pattern here is to use getOrElse

```scala
mainClass.map( runMain( _, Seq(), classLoader ) ).getOrElse( ExitCode.Failure )
```

File: stage2/Lib.scala [hunk 17]
```diff
@@ -66,6 +66,52 @@ final class Lib(logger: Logger) extends Stage1Lib(logger) with Scaffold{
...
+  
+  def mainClasses(compileTarget : File, classLoader : ClassLoader): Seq[String] = {
+    def className(file : File) = file.getName.takeWhile(_ != '.')
+    
+    def classesWithMain(target : File, classLoader : ClassLoader) : List[File]  = { 
+      def hasMainClass(file : File) : Boolean = {
+        val name = className(file)
+    
+        val inspectClass = classLoader.loadClass(name)
+        inspectClass.getDeclaredMethods().map(_.getName).contains("main")
```
Comment (cvogt) [inline stage2/Lib.scala]: this should either check if the argument list has exactly one element of type Array[String] or add a comment here that that's TODO.

File: stage2/BasicBuild.scala (hunk 13 above)
Comment (cvogt) [inline stage2/BasicBuild.scala]: the thing with headOption is, you don't need to check for the size.

File: stage2/Lib.scala [hunk 18]
```diff
@@ -66,6 +66,52 @@ final class Lib(logger: Logger) extends Stage1Lib(logger) with Scaffold{
...
+      }
+
+      def isInnerClass(file : File) : Boolean = file.toString.contains("$")
+      val files = target.listFiles
+
+      val classFiles = files.filter(file => file.toString.endsWith(".class") && !isInnerClass(file)).toList
+
+      classFiles.filter(hasMainClass(_))
```
Comment (cvogt) [inline stage2/Lib.scala]: why not return name here? If hasMainClass would be called mainClass and returns an Option[String], you could use flatMap here instead of filter and save yourself from calling .map(className(_)) again below. Also the name hasMain below is not accurate and would not be needed anymore then.

File: stage2/BasicBuild.scala (hunk 14 above)
Comment (cvogt) [inline stage2/BasicBuild.scala]: would be good if we could also add a test for this. could be one of the sample builds in there. compile it, check target exists, then clean, check it doesn't exist.

**Comment by cvogt** [comment]
would be good to get a test for this as well. Check test/ and just add builds with one main method, multiple main methods and check the results are correct. Thx Michael :).

**Comment by ChavXO** [comment]
@cvgot Moved. Working on tests.
